### ISSUE TITLE
CMakeLists defines hidden in C flags, but doesn't define LIBSSH2_API to unhide them

### ISSUE THREAD

**Issue by d3x0r** [libssh2/libssh2#1416]
**Describe the bug**
When building with CMake and GCC on Linux, I was having issues linking other code to the library that ended up containing libssh2.  I finally tracked it down to 

https://github.com/libssh2/libssh2/blob/master/CMakeLists.txt#L139-L140

``` cmake
    set(LIB_SHARED_C_FLAGS "-fvisibility=hidden")
    set(LIBSSH2_API "__attribute__ ((__visibility__ (\"default\")))")
```

that LIBSSH2_API symbol is only defined in that CMakeLists, and not used anywhere else.  The shared_C_flags ends up getting defined though so everything ends up hidden.

In the configure.ac 

           AC_DEFINE(LIBSSH2_API, [__global], [to make a symbol visible])

This would define LIBSSH2_API, but the above in cmake does not.

**To Reproduce**
Build with CMake instead of configure?  `objdump -t libssh2.so` (everything ends up with 'l' for visibility and no 'g' that would be visible externally)

I'm not sure what other options I've also forced that might also be required (I don't think there are any...).

**Comment by vszakats** [comment]
It's passed here to the code:
https://github.com/libssh2/libssh2/blob/fb6527468c26d18c800580f765ae8c8ea8edb88b/src/libssh2_config_cmake.h.in#L75-L77

Maybe some of the conditions are not correctly set or passed?

**Comment by d3x0r** [comment]
I had some time to look into this more...  I'm setting these options...

``` cmake
    SET( LIBSSH2_EXPORT_SYMBOLS ON )
    SET( LIBSSH2_BUILD_OBJECT_LIBRARY ON )
    SET( LIBSSH2_SKIP_INSTALL ON )
    set( LIBSSH2_BUILD_SHARED_LIBS OFF )
    set( LIBSSH2_BUILD_STATIC_LIBS ON )
    set( LIBSSH2_BUILD_EXAMPLES OFF )
    set( LIBSSH2_BUILD_TESTING OFF )
    set( LIBSSH2_BUILD_DOCS OFF )
```

`LIBSSH2_BUILD_OBJECT_LIBRARY` (which is a special case of `LIBSSH2_BUILD_STATIC_LIBS` )  gets the -fvisiblility=hidden compile flag, but then doesn't get `LIB_SHARED_DEFINITIONS` which includes `LIBSSH2_EXPORTS`, which is one of the flags to enable LIBSSH2_API in the configured libssh2_config.h...

Also the object library is being linked into a shared library... 

```
if(HIDE_SYMBOLS)
  set(LIB_SHARED_DEFINITIONS "LIBSSH2_EXPORTS")
```

**Comment by vszakats** [comment]
Which version of libssh2 is this happening with? Is there a CMake command-line that I could try to replicate?

These settings don't sound familiar, and grepping into the source, I cannot find them. There is `BUILD_SHARED_LIBS`, `BUILD_STATIC_LIBS`, `BUILD_EXAMPLES` and `BUILD_TESTING`, but with unprefixed names.

With SHARED=OFF and STATIC=ON, I wonder how is there a `libssh2.so` created?

curl-for-win is using a similar combo of these:
https://github.com/curl/curl-for-win/blob/330e4e814334fa87236fd9f88f357d43392f2862/libssh2.sh#L60-L63
After linking the static lib to curl, there is a post-check for symbol leaks (also on Linux), and there is none.

**Comment by d3x0r** [comment]
I see - I had to change the names - otherwise BUILD_STATIC_LIBS/BUILD_SHARED_LIBS setting affected all other libraries included.... I probably should have used external_project to include libssh2, but I just used add_subdirectory that worked with zlib,png,expat,genx,jpeg,freetype... I had looked at the git log, but apparently missed that change; must have been in an earlier merge in my fork.  Also apparently added some options because I just need the library parts.

There isn't a libssh2.so created, there's a libbag.externals.so that includes libssh2 as an object library(not static library) and all of the above other libraries...  which is why I don't need it installed and added that option. oh and object library was also something added by me... because building the .a wasn't working to link into the externals library, probably because of the hidden symbols, but also I remember it being difficult to link multiple `.a`'s into a `.so`.  Sorry I forgot I had done all of those mods.
